ReactApp/src/components/MappingMap.js

```diff
@@ -0,0 +1,168 @@
+import React, { useState, useEffect, useRef } from "react";
+import robotIconSrc from "./icons/robot.png";
+
+function MappingMap() {
+  const [mapData, setMapData] = useState(null);
+  const [robotLocation, setRobotLocation] = useState(null);
+  const [robotIcon, setRobotIcon] = useState(null);
+  const canvasRef = useRef(null);
+
+  useEffect(() => {
+    const fetchMapData = async () => {
+      try {
+        const response = await fetch(
+          "http://13.201.82.2:5747/mapping/current/map"
+        );
+        if (!response.ok) {
+          throw new Error("Network response was not ok");
+        }
+        const data = await response.json();
+        setMapData(data);
+      } catch (error) {
+        console.error("Error fetching map data:", error);
+      }
+    };
+
+    fetchMapData();
+
+    const intervalId = setInterval(fetchMapData, 5000);
+
+    return () => clearInterval(intervalId);
+  }, []);
+
+  useEffect(() => {
+    const fetchRobotLocation = async () => {
+      try {
+        const response = await fetch(
+          "http://13.201.82.2:5747/mapping/current/location"
+        );
+        if (!response.ok) {
+          throw new Error("Network response was not ok");
+        }
+        const data = await response.json();
+        setRobotLocation(data);
+        console.log(data);
+      } catch (error) {
+        console.error("Error fetching robot location:", error);
+      }
+    };
+
+    fetchRobotLocation();
+
+    const intervalId = setInterval(fetchRobotLocation, 100);
+
+    return () => clearInterval(intervalId);
+  }, []);
+
+  useEffect(() => {
+    const loadRobotIcon = () => {
+      const icon = new Image();
+      icon.src = robotIconSrc; // Update with the correct path to your icon
+      icon.onload = () => setRobotIcon(icon);
+    };
+
+    loadRobotIcon();
+  }, []);
+
+  const getColorForOccupancy = (value) => {
+    if (value === 0) return "white"; // Free space
+    if (value === 100) return "black"; // Occupied
+    return "gray"; // Unknown
+  };
+
+  const isOccupied = (value) => value === 100;
+
+  useEffect(() => {
+    const drawRobot = (ctx, location, mapInfo, cellWidth, cellHeight) => {
+      const { x, y, theta } = location;
+      // const robotX = (x - mapInfo.origin.position.x) / mapInfo.resolution * cellWidth;
+      // const robotY = (y - mapInfo.origin.position.y) / mapInfo.resolution * cellHeight;
+      const robotX =
+        ((-x - mapInfo.origin.position.x) / mapInfo.resolution) * cellWidth;
+      const robotY =
+        ((y - mapInfo.origin.position.y) / mapInfo.resolution) * cellHeight;
```

Comment:
Look for the initial position error problem, debug using ros topics or cub scripts for data locally or may use opencv with imageto draw robot and test

Reply:
solved, changed to new and correct equations